<a href="https://colab.research.google.com/github/wanutchapornmun/229352-StatisticalLearning/blob/main/Lab04_Naive_Bayes_Grid_and_Random_Search_660510561.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Statistical Learning for Data Science 2 (229352)
#### Instructor: Donlapark Ponnoprat

#### [Course website](https://donlapark.pages.dev/229352/)

## Lab #4

In [9]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

from scipy.stats import uniform

In [10]:
train = fetch_20newsgroups(subset='train')
test = fetch_20newsgroups(subset='test')

Xtrain = train.data[:3000]
ytrain = train.target[:3000]
Xtest = test.data[:500]
ytest = test.target[:500]

print("X:", len(Xtest))
print("y:", len(ytest))

X: 500
y: 500


### Naive Bayes [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html)

In [11]:
pipeline = Pipeline([('tfidf', TfidfVectorizer(stop_words = 'english')),
                     ('nb', MultinomialNB())])

pipeline.fit(Xtrain, ytrain)

# เมื่อ fit เสร็จ ก็เอาไปทำนายบน test set
ypred = pipeline.predict(Xtest)

print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

           0       0.67      0.38      0.48        21
           1       0.79      0.52      0.63        21
           2       0.58      0.69      0.63        26
           3       0.74      0.68      0.71        34
           4       0.72      0.85      0.78        34
           5       0.88      0.81      0.84        26
           6       1.00      0.73      0.84        22
           7       0.70      1.00      0.82        28
           8       0.90      0.82      0.86        33
           9       0.88      0.84      0.86        25
          10       0.82      1.00      0.90        27
          11       0.79      0.95      0.86        20
          12       0.59      0.54      0.57        24
          13       0.75      0.78      0.77        23
          14       0.87      0.71      0.78        28
          15       0.53      0.90      0.67        29
          16       0.50      0.95      0.66        21
          17       0.94    

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


### Random Search Cross-Validation [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)

### Uniform distribution in `Scipy` [(Documentation)](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.uniform.html)

In [12]:
# การ search หา พารามิเตอร์ที่เหมาะสม
parameters = {'nb__alpha': uniform(loc=0, scale=10)}

clf = RandomizedSearchCV(pipeline, parameters, n_iter=7)

clf.fit(Xtrain, ytrain) # ผลลัพธ์ที่ได้จะเป็นโมเดลที่ดีที่สุด

RandomizedSearchCV(estimator=Pipeline(steps=[('tfidf',
                                              TfidfVectorizer(stop_words='english')),
                                             ('nb', MultinomialNB())]),
                   n_iter=7,
                   param_distributions={'nb__alpha': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7aaec6bb2b10>})

In [13]:
ypred = clf.predict(Xtest)

print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

           0       0.73      0.38      0.50        21
           1       0.90      0.43      0.58        21
           2       0.50      0.73      0.59        26
           3       0.73      0.56      0.63        34
           4       0.64      0.85      0.73        34
           5       0.81      0.81      0.81        26
           6       1.00      0.64      0.78        22
           7       0.57      1.00      0.73        28
           8       0.90      0.82      0.86        33
           9       0.87      0.80      0.83        25
          10       0.79      1.00      0.89        27
          11       0.83      0.95      0.88        20
          12       0.59      0.54      0.57        24
          13       0.80      0.70      0.74        23
          14       0.89      0.57      0.70        28
          15       0.50      0.90      0.64        29
          16       0.49      0.90      0.63        21
          17       0.93    

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


#### Exercise

1. For the Naive Bayes model, use grid search 5-fold cross-validation across different values of `alpha` to find the best model.

2. For the best value of `alpha`, compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* What is the model's `f1_macro` score?

3. Repeat Exercise 1 and 2 for **random search** 5-fold cross validation across different values of `alpha`. Compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* Did you get a better `f1_macro` score compared to grid search in Exercise 2?

ทำ 5-fold ให้ cv = 5

ขั้นตอนสุดท้ายบอกด้วยว่า ระหว่าง grid search และ randomize ตัวไหนให้ค่าพารามิเตอร์ที่ดีที่สุด

In [14]:
from sklearn.neighbors import KNeighborsClassifier

In [19]:
pipeline_grid = Pipeline([('tfidf', TfidfVectorizer(stop_words='english')),
 ('nb', MultinomialNB())])

parameters_grid = {'nb__alpha': [0.01, 0.1, 0.5, 1.0, 5.0]}

grid = GridSearchCV(pipeline_grid, parameters_grid,
                      scoring='f1_macro', cv=5)

grid.fit(Xtrain, ytrain)

best_alpha_grid = grid.best_params_['nb__alpha']
print("Best alpha (Grid Search):", best_alpha_grid)

Best alpha (Grid Search): 0.01


In [22]:
from sklearn.metrics import f1_score

In [25]:
y_pred_grid = grid.predict(Xtest)

f1_grid = f1_score(ytest, y_pred_grid, average='macro')

print(classification_report(ytest, y_pred_grid))
print("f1_macro (Grid Search):", f1_grid)

              precision    recall  f1-score   support

           0       0.80      0.57      0.67        21
           1       0.57      0.62      0.59        21
           2       0.62      0.58      0.60        26
           3       0.75      0.71      0.73        34
           4       0.84      0.79      0.82        34
           5       0.89      0.62      0.73        26
           6       0.78      0.82      0.80        22
           7       0.88      1.00      0.93        28
           8       0.97      0.88      0.92        33
           9       0.88      0.92      0.90        25
          10       0.87      0.96      0.91        27
          11       0.83      0.95      0.88        20
          12       0.67      0.75      0.71        24
          13       0.83      0.87      0.85        23
          14       0.81      0.93      0.87        28
          15       0.75      0.93      0.83        29
          16       0.53      0.90      0.67        21
          17       0.94    

In [42]:
pipeline_random_search = Pipeline([('tfidf', TfidfVectorizer(stop_words='english')),
    ('nb', MultinomialNB())
])

parameters_random_search = {'nb__alpha': uniform(loc=0, scale=10)}

clf_random_search = RandomizedSearchCV(pipeline_random_search,
                                       parameters_random_search, n_iter=15,
                                       scoring='f1_macro', cv=5)

clf_random_search.fit(Xtrain, ytrain)


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('tfidf',
                                              TfidfVectorizer(stop_words='english')),
                                             ('nb', MultinomialNB())]),
                   n_iter=15,
                   param_distributions={'nb__alpha': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7aaec6100200>},
                   scoring='f1_macro')

In [43]:
ypred_random_search = clf_random_search.predict(Xtest)

print(classification_report(ytest, ypred_random_search))

              precision    recall  f1-score   support

           0       0.73      0.52      0.61        21
           1       0.68      0.62      0.65        21
           2       0.60      0.58      0.59        26
           3       0.71      0.71      0.71        34
           4       0.88      0.85      0.87        34
           5       0.89      0.62      0.73        26
           6       0.94      0.77      0.85        22
           7       0.75      0.96      0.84        28
           8       0.97      0.85      0.90        33
           9       0.92      0.88      0.90        25
          10       0.84      1.00      0.92        27
          11       0.77      1.00      0.87        20
          12       0.59      0.71      0.64        24
          13       0.83      0.83      0.83        23
          14       0.86      0.86      0.86        28
          15       0.59      0.93      0.72        29
          16       0.51      0.95      0.67        21
          17       0.94    

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [44]:
print("Best alpha (Random Search):", clf_random_search.best_params_['nb__alpha'])

Best alpha (Random Search): 0.17144136665418808


In [45]:
ypred_random_search = clf_random_search.predict(Xtest)

print(classification_report(ytest, ypred_random_search))

f1_random_search = f1_score(ytest, ypred_random_search, average='macro')

              precision    recall  f1-score   support

           0       0.73      0.52      0.61        21
           1       0.68      0.62      0.65        21
           2       0.60      0.58      0.59        26
           3       0.71      0.71      0.71        34
           4       0.88      0.85      0.87        34
           5       0.89      0.62      0.73        26
           6       0.94      0.77      0.85        22
           7       0.75      0.96      0.84        28
           8       0.97      0.85      0.90        33
           9       0.92      0.88      0.90        25
          10       0.84      1.00      0.92        27
          11       0.77      1.00      0.87        20
          12       0.59      0.71      0.64        24
          13       0.83      0.83      0.83        23
          14       0.86      0.86      0.86        28
          15       0.59      0.93      0.72        29
          16       0.51      0.95      0.67        21
          17       0.94    

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [47]:
print("f1_macro (Random Search):", f1_random_search)

f1_macro (Random Search): 0.7255381028880163


# **วิธี Grid Search**
alpha = 0.01\
f1_macro = 0.7708784080844321

# **วิธี Random Search**
alpha = 0.17144136665418808\
f1_macro = 0.7255381028880163

# Grid Search VS Random Search
**Ans** : Grid Search เพราะ ให้ค่า f1_macro สูงกว่า Random Search


